core/src/main/java/org/elasticsearch/rest/action/RestBuilderListener.java

```diff
@@ -28,19 +28,25 @@
  */
 public abstract class RestBuilderListener<Response> extends RestResponseListener<Response> {
 
+    // pkg-private static boolean that enables testing of the auto-close functionality
+    static boolean assertionsEnabled = true;
+
     public RestBuilderListener(RestChannel channel) {
         super(channel);
     }
 
     @Override
     public final RestResponse buildResponse(Response response) throws Exception {
-        try (XContentBuilder builder = channel.newBuilder()){
-            return buildResponse(response, builder);
+        try (XContentBuilder builder = channel.newBuilder()) {
+            final RestResponse restResponse = buildResponse(response, builder);
+            assert assertionsEnabled == false || builder.isClosed() : "callers should ensure the XContentBuilder is closed themselves";
```

Comment:
can you add a protected method `assert assertBuilderClosed(builder)` we can override in tests instead. this is very confusing and I hate static mutable vars